### PR TITLE
CLN refactor core/arrays

### PR DESCRIPTION
**Pull request by MarcoGorelli**
Some refactorings found by Sourcery https://sourcery.ai/

I've removed the ones of the kind
```diff
- if param:
-     var = a
- else:
-     var = b
+ var = a if param else b
```

### CODE REVIEW

File: pandas/core/arrays/boolean.py [hunk 1]
```diff
@@ -170,12 +170,11 @@ def coerce_to_array(
         values[~mask_values] = values_object[~mask_values].astype(bool)
 
         # if the values were integer-like, validate it were actually 0/1's
-        if inferred_dtype in integer_like:
-            if not np.all(
-                values[~mask_values].astype(float)
-                == values_object[~mask_values].astype(float)
-            ):
-                raise TypeError("Need to pass bool-like values")
+        if inferred_dtype in integer_like and not np.all(
```
Comment (jreback) [inline pandas/core/arrays/boolean.py:173]: would use parens here to make this easier to read / line up

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/sparse/array.py [hunk 2]
```diff
@@ -736,25 +734,17 @@ def value_counts(self, dropna=True):
 
         keys, counts = algos.value_counts_arraylike(self.sp_values, dropna=dropna)
         fcounts = self.sp_index.ngaps
-        if fcounts > 0:
-            if self._null_fill_value and dropna:
-                pass
+        if fcounts > 0 and (not self._null_fill_value or not dropna):
```
Comment (jreback) [inline pandas/core/arrays/sparse/array.py:737]: this is nice but it may have changed the semantics

**Comment by jreback** [comment]
this is failing checks

**Comment by MarcoGorelli** [comment]
> this is failing checks

Yes, the change
```diff
-        for i in range(len(self)):
-            yield self[i]
+        yield from self
```
was throwing a RecursionError

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @MarcoGorelli

**Pull request merged**